tests/Api/UploadTest.php

```diff
@@ -0,0 +1,418 @@
+					'Upload-Offset' => 0,
+					'Upload-Id'     => 'abcd'
+				]
+			]
+		]);
+
+		$dir = static::TMP . '/site/cache/.uploads';
+		$this->assertNull(Upload::chunk($api, $source, basename($source)));
+		$this->assertFileDoesNotExist($dir . '/test.md');
+		$this->assertFileExists($dir . '/abcd-test.md');
+	}
+
+	/**
+	 * @covers ::chunk
+	 */
+	public function testChunkIdRemoveUnallowedCharacters()
+	{
+		$source = static::TMP . '/test.md';
+		F::write($source, 'abcdef');
+
+		$api = new Api([
+			'requestData' => [
+				'headers' => [
+					'Upload-Length' => 3000,
+					'Upload-Offset' => 0,
+					'Upload-Id'     => '../a/b!!cd'
+				]
+			]
+		]);
+
+		$dir = static::TMP . '/site/cache/.uploads';
+		$this->assertNull(Upload::chunk($api, $source, basename($source)));
+		$this->assertFileDoesNotExist($dir . '/test.md');
+		$this->assertFileExists($dir . '/abcd-test.md');
+	}
+
+	/**
+	 * @covers ::chunk
+	 */
+	public function testChunkFilenameNoDirectoryTraversal()
+	{
+		$source = static::TMP . '/test.md';
+		F::write($source, 'abcdef');
+
+		$api = new Api([
+			'requestData' => [
+				'headers' => [
+					'Upload-Length' => 3000,
+					'Upload-Offset' => 0,
+					'Upload-Id'     => 'abcd'
+				]
+			]
+		]);
+
+		$dir = static::TMP . '/site/cache/.uploads';
+		$this->assertNull(Upload::chunk($api, $source, '../../test.md'));
+		$this->assertFileDoesNotExist($dir . '/test.md');
+		$this->assertFileExists($dir . '/abcd-test.md');
+	}
+
+	/**
+	 * @covers ::chunk
+	 */
+	public function testChunkSuccesfulAll()
+	{
+		$source = static::TMP . '/test.md';
+		F::write($source, 'abc');
+		$api    = new Api([
+			'requestData' => [
+				'headers' => [
+					'Upload-Length' => 6,
+					'Upload-Offset' => 0,
+					'Upload-Id'     => 'abcd'
+				]
+			]
+		]);
+
+		$this->assertNull(Upload::chunk($api, $source, basename($source)));
+
+		$source = static::TMP . '/test.md';
+		F::write($source, 'def');
+		$api    = new Api([
+			'requestData' => [
+				'headers' => [
+					'Upload-Length' => 6,
+					'Upload-Offset' => 3,
+					'Upload-Id'     => 'abcd'
+				]
+			]
+		]);
+
+		$file = static::TMP . '/site/cache/.uploads/test.md';
+		$this->assertSame($file, Upload::chunk($api, $source, 'test.md'));
+		$this->assertFileExists($file);
+		$this->assertFileDoesNotExist('abcd-' . $file);
+		$this->assertSame('abcdef', F::read($file));
+	}
+
+	/**
+	 * @covers ::chunkSize
+	 */
+	public function testChunkSize()
+	{
+		ini_set('upload_max_filesize', '10M');
+		ini_set('post_max_size', '20M');
+
+		$this->assertSame(
+			(int)floor(10 * 1024 * 1024 * 0.95),
+			Upload::chunkSize()
+		);
+
+		// with CloudFlare
+		ini_set('upload_max_filesize', '200M');
+		ini_set('post_max_size', '200M');
+		$_SERVER['HTTP_CF_CONNECTING_IP'] = '1.1.1.1';
+
+		$this->assertSame(
+			(int)floor(100 * 1024 * 1024 * 0.95),
+			Upload::chunkSize()
+		);
+
+		ini_restore('upload_max_filesize');
+		ini_restore('post_max_size');
+		unset($_SERVER['HTTP_CF_CONNECTING_IP']);
```

Comment:
Those should be in `tearDown()`, otherwise they won't be reset if the test fails, thus changing the state for following tests.